File: Code/Go/Challenge012.go
package main

import(
	"os"
	"fmt"
	"io/ioutil"
	"bytes"
)

func (c *Challenge) Challenge012() {
	path:=".\\Data\\012"
	if !PathExists(path) {
		err := os.Mkdir(path, os.ModeDir)
		if err != nil {
			fmt.Printf("mkdir failed![%v]\n", err)
			return
		}
	}

	gfx:="http://www.pythonchallenge.com/pc/return/evil2.gfx"
	filename:=path+"\\evil2.gfx"
	DownloadWithBasicAuth(gfx, filename, "huge", "file")

	evil4:="http://www.pythonchallenge.com/pc/return/evil4.jpg"
	DownloadWithBasicAuth(evil4, path+"\\evil4.jpg", "huge", "file")

	cnt, err := ioutil.ReadFile(filename)
	if err != nil {
		fmt.Printf("read file failed![%v]\n", err)
		return
	}

	exts :=[]string {"jpg","png","gif","png","jpg"}
	for i := 0; i < 5; i++ {
		outfile := fmt.Sprintf("%s\\out%d.%s", path, i, exts[i])
		var builder bytes.Buffer
		for j := i; j < len(cnt); j+=5 {
			builder.WriteByte(cnt[j])
		}
		err := ioutil.WriteFile(outfile, builder.Bytes(), 0666)
		if err != nil {
			fmt.Printf("write file failed![%v]\n", err)
			return
		}
	}
}

File: Code/Python/Challenge012.py
import helper
path=".\\Data\\012"
helper.ensureDir(path)

# ================================
# 使用http认证，下载文件
import urllib.request
helper.installHTTPBasicAuthOpener("huge", "file")

gfx="http://www.pythonchallenge.com/pc/return/evil2.gfx"
(filename, headers)=urllib.request.urlretrieve(gfx, path+"\\evil2.gfx")

evil4="http://www.pythonchallenge.com/pc/return/evil4.jpg"
urllib.request.urlretrieve(evil4, path+"\\evil4.jpg")
# ================================

filename=path+"\\evil2.gfx"
# ================================
# 输出文件头
# fp=open(filename,'rb')
# cnt=fp.read(16*7-2)
# fp.close()

# for i in range(5):
#     print(cnt[i::5])
# ================================

fp=open(filename,'rb')
cnt=fp.read()
fp.close()

exts=["jpg","png","gif","png","jpg"]
for i in range(5):
    fp=open(path+"\\out%d.%s" % (i, exts[i]),'wb')
    fp.write(cnt[i::5])
    fp.close()

print("please see the folder: ", path)